### PR TITLE
Saas 12.2 elearning fix completion tde

### PR DESCRIPTION
**Pull request by tde-banana-odoo**

### CODE REVIEW

File: addons/website_slides/models/slide_channel.py [hunk 1]
```diff
@@ -35,29 +35,32 @@ def _compute_completion(self):
...
+        to_complete = self.env['slide.channel.partner']
+        if values.get('completion'):
+            incomplete_self = self.filtered(lambda cp: not cp.completed)
+            channels_data = {result['id']: result['total_slides'] for result in incomplete_self.mapped('channel_id').read(['total_slides'])}
+            for cp in incomplete_self:
+                if values.get('completion') >= channels_data[cp.channel_id.id]:
+                    to_complete |= cp
+
+            partner_karma = dict.fromkeys(incomplete_self.mapped('partner_id').ids, 0)
+            for channel_partner in incomplete_self:
```
Comment (awa-odoo) [inline addons/website_slides/models/slide_channel.py]: If I'm not mistaken you add karma for all incomplete channels, which doesn't seem right.

Comment (tde-banana-odoo) [reply addons/website_slides/models/slide_channel.py]: It seems you were right. I added some tests and fixed the issue.

File: addons/website_slides/models/slide_slide.py [hunk 2]
```diff
@@ -407,22 +407,25 @@ def _action_vote(self, upvote=True):
         slide_id = slide_partners.mapped('slide_id')
         new_slides = self_sudo - slide_id
         channel = slide_id.channel_id
+        karma_to_add = 0
 
         for slide_partner in slide_partners:
             if upvote:
```
Comment (awa-odoo) [inline addons/website_slides/models/slide_slide.py]: Maybe 

```
if upvote and slide_partner.vote != -1:
    new_vote = 1
elif not upvote and slide_partner.vote != 1:
    new_vote = -1
else:
    new_vote = 0
```

Would be easier to read?

Comment (tde-banana-odoo) [reply addons/website_slides/models/slide_slide.py]: I added some code to protect karma addition against upvoting an already-upvoted or downvoted slide. As I added code in this if / else I would prefer to keep the current structure. I let you look at the final code.

Comment (tde-banana-odoo) [reply addons/website_slides/models/slide_slide.py]: Just rewrite it. Maybe your structure will be better finally.

Do you love spam ?

Comment (tde-banana-odoo) [reply addons/website_slides/models/slide_slide.py]: Nope, I keep the current one.

Comment (tde-banana-odoo) [reply addons/website_slides/models/slide_slide.py]: What's in the box ?

Comment (awa-odoo) [reply addons/website_slides/models/slide_slide.py]: Gyproc ?

Comment (tde-banana-odoo) [reply addons/website_slides/models/slide_slide.py]: Gyproc karma !

Comment (jem-odoo) [reply addons/website_slides/models/slide_slide.py]: https://www.youtube.com/watch?v=EuRKX-LrzDQ

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-ff

File: addons/website_slides/models/slide_channel.py [hunk 3]
```diff
@@ -35,29 +36,32 @@ def _compute_completion(self):
...
-            incomplete_channel_partners = self.filtered(lambda cp: not cp.completed)
-            partner_karma = dict.fromkeys(incomplete_channel_partners.mapped('partner_id').ids, 0)
-            for channel_partner in incomplete_channel_partners:
+        to_complete = self.env['slide.channel.partner']
+        if values.get('completion'):
+            incomplete_self = self.filtered(lambda cp: not cp.completed)
+            channels_data = {result['id']: result['total_slides'] for result in incomplete_self.mapped('channel_id').read(['total_slides'])}
+            for cp in incomplete_self:
+                if values.get('completion') >= channels_data[cp.channel_id.id]:
+                    to_complete |= cp
```
Comment (dbeguin) [inline addons/website_slides/models/slide_channel.py]: Couldn't we reduce this a bit ?
```
to_complete_ids = [result['id'] for result in incomplete_self.mapped('channel_id').read(['total_slides']) 
                               if values.get('completion') >= result['total_slides']]
to_complete = self.filtered(lambda cp: cp.id in to_complete_ids) (or browse ?)
```

Comment (tde-banana-odoo) [reply addons/website_slides/models/slide_channel.py]: It is never too late.

Comment (tde-banana-odoo) [reply addons/website_slides/models/slide_channel.py]: Well actually in your code to_complete_ids is a ID list of channels. We could have to_complete = self.filtered(lambda cp: cp.channel_id.id in to_complete_ids) . Anyway I think it is globally equivalent as we don't gain a lot of lines.

Comment (dbeguin) [reply addons/website_slides/models/slide_channel.py]: ah indeed ! but well, it was the idea :) bisous!

**Pull request closed without merging**